builtin/provisioners/salt-masterless/resource_provisioner.go

```diff
@@ -0,0 +1,468 @@
+	if err := p.createDir(o, comm, p.TempConfigDir); err != nil {
+		return fmt.Errorf("Error creating remote temporary directory: %s", err)
+	}
+
+	if p.MinionConfig != "" {
+		o.Output(fmt.Sprintf("Uploading minion config: %s", p.MinionConfig))
+		src = p.MinionConfig
+		dst = filepath.ToSlash(filepath.Join(p.TempConfigDir, "minion"))
+		if err = p.uploadFile(o, comm, dst, src); err != nil {
+			return fmt.Errorf("Error uploading local minion config file to remote: %s", err)
+		}
+
+		// move minion config into /etc/salt
+		o.Output(fmt.Sprintf("Make sure directory %s exists", "/etc/salt"))
+		if err := p.createDir(o, comm, "/etc/salt"); err != nil {
+			return fmt.Errorf("Error creating remote salt configuration directory: %s", err)
+		}
+		src = filepath.ToSlash(filepath.Join(p.TempConfigDir, "minion"))
+		dst = "/etc/salt/minion"
+		if err = p.moveFile(o, comm, dst, src); err != nil {
+			return fmt.Errorf("Unable to move %s/minion to /etc/salt/minion: %s", p.TempConfigDir, err)
+		}
+	}
+
+	o.Output(fmt.Sprintf("Uploading local state tree: %s", p.LocalStateTree))
+	src = p.LocalStateTree
+	dst = filepath.ToSlash(filepath.Join(p.TempConfigDir, "states"))
+	if err = p.uploadDir(o, comm, dst, src, []string{".git"}); err != nil {
+		return fmt.Errorf("Error uploading local state tree to remote: %s", err)
+	}
+
+	// move state tree from temporary directory
+	src = filepath.ToSlash(filepath.Join(p.TempConfigDir, "states"))
+	dst = p.RemoteStateTree
+	if err = p.removeDir(o, comm, dst); err != nil {
+		return fmt.Errorf("Unable to clear salt tree: %s", err)
+	}
+	if err = p.moveFile(o, comm, dst, src); err != nil {
+		return fmt.Errorf("Unable to move %s/states to %s: %s", p.TempConfigDir, dst, err)
+	}
+
+	if p.LocalPillarRoots != "" {
+		o.Output(fmt.Sprintf("Uploading local pillar roots: %s", p.LocalPillarRoots))
+		src = p.LocalPillarRoots
+		dst = filepath.ToSlash(filepath.Join(p.TempConfigDir, "pillar"))
+		if err = p.uploadDir(o, comm, dst, src, []string{".git"}); err != nil {
+			return fmt.Errorf("Error uploading local pillar roots to remote: %s", err)
+		}
+
+		// move pillar root from temporary directory
+		src = filepath.ToSlash(filepath.Join(p.TempConfigDir, "pillar"))
+		dst = p.RemotePillarRoots
+
+		if err = p.removeDir(o, comm, dst); err != nil {
+			return fmt.Errorf("Unable to clear pillar root: %s", err)
+		}
+		if err = p.moveFile(o, comm, dst, src); err != nil {
+			return fmt.Errorf("Unable to move %s/pillar to %s: %s", p.TempConfigDir, dst, err)
+		}
+	}
+
+	o.Output(fmt.Sprintf("Running: salt-call --local %s", p.CmdArgs))
+	cmd := &remote.Cmd{Command: p.sudo(fmt.Sprintf("salt-call --local %s", p.CmdArgs))}
+	if err = comm.Start(cmd); err != nil || cmd.ExitStatus != 0 {
+		if err == nil {
+			err = fmt.Errorf("Bad exit status: %d", cmd.ExitStatus)
+		}
+
+		return fmt.Errorf("Error executing salt-call: %s", err)
+	}
+
+	return nil
+}
+
+func (p *provisioner) Cancel() {
```

Comment:
This can be removed, as it doesn't apply to the terraform provisioner.

we can work later on making this "stoppable", using `comm.Disconnect` so the provisioner in a controlled manner when `Stop()` is called on the provisioner.